dm/master/openapi.go

```diff
@@ -312,3 +376,201 @@ func modelToSourceCfg(source openapi.Source) *config.SourceConfig {
+		}
+	}
+	// start to generate sub task configs
+	subTaskCfgList := make([]config.SubTaskConfig, len(task.SourceConfig.SourceConf))
+	for i, sourceCfg := range task.SourceConfig.SourceConf {
+		subTaskCfg := config.NewSubTaskConfig()
+		// set task name and mode
+		subTaskCfg.Name = task.Name
+		subTaskCfg.Mode = string(task.TaskMode)
+		// set task meta
+		subTaskCfg.MetaFile = *task.MetaSchema
+		if meta, ok := sourceDBMetaMap[sourceCfg.SourceName]; ok {
+			subTaskCfg.Meta = meta
+		}
+		// set shard config
+		if task.ShardMode != nil {
+			subTaskCfg.IsSharding = true
+			mode := *task.ShardMode
+			subTaskCfg.ShardMode = string(mode)
+		} else {
+			subTaskCfg.IsSharding = false
+		}
+		// set online ddl pulgin config
+		subTaskCfg.OnlineDDL = task.EnhanceOnlineSchemaChange
+		// set case sensitive from source
+		subTaskCfg.CaseSensitive = sourceCfgMap[sourceCfg.SourceName].CaseSensitive
+		// set source db config
+		subTaskCfg.SourceID = sourceCfg.SourceName
+		subTaskCfg.From = sourceCfgMap[sourceCfg.SourceName].From
+		// set target db config
+		subTaskCfg.To = *toDBCfg.Clone()
+		// TODO set meet error policy
+		// TODO ExprFilter
+		// set full unit config
+		subTaskCfg.MydumperConfig = config.DefaultMydumperConfig()
+		subTaskCfg.LoaderConfig = config.DefaultLoaderConfig()
+		if fullCfg := task.SourceConfig.FullMigrateConf; fullCfg != nil {
+			if fullCfg.ExportThreads != nil {
+				subTaskCfg.MydumperConfig.Threads = *fullCfg.ExportThreads
+			}
+			if fullCfg.DataDir != nil {
+				subTaskCfg.LoaderConfig.Dir = *fullCfg.DataDir
+			}
+		}
+		// set incremental config
+		subTaskCfg.SyncerConfig = config.DefaultSyncerConfig()
+		if incrCfg := task.SourceConfig.IncrMigrateConf; incrCfg != nil {
+			if incrCfg.ReplThreads != nil {
+				subTaskCfg.SyncerConfig.WorkerCount = *incrCfg.ReplThreads
+			}
+			if incrCfg.ReplBatch != nil {
+				subTaskCfg.SyncerConfig.Batch = *incrCfg.ReplBatch
+			}
+		}
+		// set route , blockAllowList, filter config
+		doDBs := []string{}
+		doTables := []*filter.Table{}
+		routeRules := []*router.TableRule{}
+		filterRules := []*bf.BinlogEventRule{}
+		for _, rule := range tableMigrateRuleMap[sourceCfg.SourceName] {
+			// route
+			routeRules = append(routeRules, &router.TableRule{
+				SchemaPattern: rule.Source.Schema,
+				TablePattern:  rule.Source.Table,
+				TargetSchema:  rule.Target.Schema,
+				TargetTable:   rule.Target.Table,
+			})
+			// filter
+			if rule.EventFilterName != nil {
+				for _, name := range *rule.EventFilterName {
+					filterRule, ok := eventFilterTemplateMap[name] // NOTE: this return a cpoied value
+					if !ok {
+						return nil, terror.ErrOpenAPICommonError.Generatef("filter rule name=%s not found", name)
+					}
+					filterRule.SchemaPattern = rule.Source.Schema
+					filterRule.TablePattern = rule.Source.Table
+					filterRules = append(filterRules, &filterRule)
+				}
+			}
+			// BlockAllowList
+			doDBs = append(doDBs, rule.Source.Schema)
```

Comment:
duplicated value

Reply:
fixed in L541 `subTaskCfg.BAList = &filter.Rules{DoDBs: removeDuplication(doDBs), DoTables: doTables`